generic/piranha-tree-sitter/src/piranha.rs

```diff
@@ -0,0 +1,534 @@
+  /// The `edit:InputEdit` performed.
+  ///
+  /// Note - Causes side effect. - Updates `self.ast` and `self.code`
+  fn apply_edit(
+    &mut self, replace_range: Range, replacement_str: String, parser: &mut Parser,
+  ) -> InputEdit {
+    // Get the tree_sitter's input edit representation
+    let (new_source_code, edit) = self.get_edit(replace_range, &replacement_str);
+    // Apply edit to the tree
+    self.ast.edit(&edit);
+    // Create a new updated tree from the previous tree
+    let new_tree = parser
+      .parse(&new_source_code, Some(&self.ast))
+      .expect("Could not generate new tree!");
+    self.ast = new_tree;
+    self.code = new_source_code;
+    return edit;
+  }
+
+  /// Will apply the `rule` to all of its occurrences in the source code unit.
+  fn apply_rule(
+    &mut self, rule: Rule, rules_store: &mut RuleStore, parser: &mut Parser,
+    scope_query: &Option<String>,
+  ) {
+    loop {
+      if !self.apply_rule_to_first_match(rule.clone(), rules_store, parser, scope_query) {
+        break;
+      }
+    }
+  }
+
+  // Applies the rule to the first match in the source code
+  fn apply_rule_to_first_match(
+    &mut self, rule: Rule, rules_store: &mut RuleStore, parser: &mut Parser,
+    scope_query: &Option<String>,
+  ) -> bool {
+    // Get scope node
+    let mut scope_node = self.ast.root_node();
+    if let Some(query_str) = scope_query {
+      // Apply the scope query in the source code and get the appropriate node
+      let tree_sitter_scope_query = rules_store.get_query(query_str);
+      if let Some((range, _)) =
+        &self
+          .ast
+          .root_node()
+          .get_match_for_query(&self.code, tree_sitter_scope_query, true)
+      {
+        scope_node = self.get_node_for_range(range.start_byte, range.end_byte);
+      }
+    }
+
+    let mut any_match = false;
+
+    // Match the rule "anywhere" inside the scope_node
+    if let Some((range, rpl, code_snippets_by_tag)) =
+      self.get_any_match_for_rule(&rule, rules_store, scope_node, true)
+    {
+      any_match = true;
+      // Get the edit for applying the matched rule to the source code
+      let edit = self.apply_edit(range, rpl, parser);
+
+      // Add all the (code_snippet, tag) mapping to the substitution table.
+      self.substitutions.extend(code_snippets_by_tag);
+
+      let mut current_edit = edit.clone();
+      let mut current_rule = rule.clone();
+      let mut file_level_next_rules = vec![];
+
+      // perform the parent edits, while queueing the Method and Class level edits.
+      let file_level_scope_names = [METHOD, CLASS];
+      loop {
+        // Get all the (next) rules that could be after applying the current rule (`rule`)
+        let next_rules_by_scope = rules_store.get_next(&current_rule, &self.substitutions);
+        for (scope_level, rules) in &next_rules_by_scope {
+          // Scope level will be "Class" or "Method"
+          if file_level_scope_names.contains(&scope_level.as_str()) {
```

Comment:
Is there a enum type that can be used here instead of string comparisons?

Reply:
I was thinking not to hardcode it as an enum so that scope generation is more flexible. For instance, someone could add a scope generator for a lambda expression if they want to.